pkg/exchange/coinbase/exchage.go

```diff
@@ -0,0 +1,309 @@
+	ticker, err := req.Do(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get ticker: %v", symbol)
+	}
+	return toGlobalTicker(ticker), nil
+}
+
+func (e *Exchange) QueryTickers(ctx context.Context, symbol ...string) (map[string]types.Ticker, error) {
+	tickers := make(map[string]types.Ticker)
+	for _, s := range symbol {
+		ticker, err := e.QueryTicker(ctx, s)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get tickers")
+		}
+		tickers[s] = *ticker
+	}
+	return tickers, nil
+}
+
+func (e *Exchange) QueryKLines(ctx context.Context, symbol string, interval types.Interval, options types.KLineQueryOptions) ([]types.KLine, error) {
+	if !e.IsSupportedInterval(interval) {
+		return nil, errors.Errorf("unsupported interval: %v", interval)
+	}
+	// default limit is 300, which is the maximum limit of the Coinbase Exchange API
+	if options.Limit == 0 {
+		options.Limit = 300
+	}
+	if options.Limit > 300 {
+		log.Warnf("limit %d is greater than the maximum limit 300, set to 300", options.Limit)
+		options.Limit = 300
+	}
+	granity := interval.String()
+	req := e.client.NewGetCandlesRequest().ProductID(toLocalSymbol(symbol)).Granularity(granity)
+	if options.StartTime != nil {
+		req.Start(*options.StartTime)
+	}
+	if options.EndTime != nil {
+		req.End(*options.EndTime)
+	}
+	candles, err := req.Do(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get klines(%v): %v", interval, symbol)
+	}
+	if len(candles) > options.Limit {
+		candles = candles[:options.Limit]
+	}
+	klines := make([]types.KLine, 0, len(candles))
+	for _, candle := range candles {
+		klines = append(klines, *toGlobalKline(symbol, granity, &candle))
+	}
+	return klines, nil
+}
+
+// ExchangeOrderQueryService
+func (e *Exchange) QueryOrder(ctx context.Context, q types.OrderQuery) (*types.Order, error) {
+	req := e.client.NewSingleOrderRequst().OrderID(q.OrderID)
+	cbOrder, err := req.Do(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get order: %v", q.OrderID)
+	}
+	return toGlobalOrder(cbOrder), nil
+}
+
+func (e *Exchange) QueryOrderTrades(ctx context.Context, q types.OrderQuery) ([]types.Trade, error) {
+	cbTrades, err := e.queryOrderTradesByPagination(ctx, q.OrderID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get order trades: %v", q.OrderID)
+	}
+	trades := make([]types.Trade, 0, len(cbTrades))
+	for _, cbTrade := range cbTrades {
+		trades = append(trades, *toGlobalTrade(&cbTrade))
+	}
+	return trades, nil
+}
+
+func (e *Exchange) queryOrderTradesByPagination(ctx context.Context, orderID string) (api.TradeSnapshot, error) {
+	paginationLimit := 100
+	req := e.client.NewGetOrderTradesRequest().OrderID(orderID).Limit(paginationLimit)
+	cbTrades, err := req.Do(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if len(cbTrades) < paginationLimit {
+		return cbTrades, nil
+	}
+
+	if len(cbTrades) < paginationLimit {
+		return cbTrades, nil
+	}
+	done := false
+	for {
+		if done {
```

Comment:
you need to handle the ctx.Done() in the loop.

Reply:
Done.